graphs/minimum_spanning_tree_kruskal2.py

```diff
@@ -0,0 +1,111 @@
+from __future__ import annotations
+
+
+class Disjoint_Set_Tree_Node:
+    # Disjoint Set Node to store the parent and rank
+    def __init__(self, key: int) -> None:
+        self.key = key
+        self.parent = self
+        self.rank = 0
+
+
+class Disjoint_Set_Tree:
+    # Disjoint Set DataStructure
+    def __init__(self):
+        # map from node name to the node object
+        self.map = {}
+
+    def make_set(self, x: int) -> None:
+        # create a new set with x as its member
+        self.map[x] = Disjoint_Set_Tree_Node(x)
+
+    def find_set(self, x: int) -> Disjoint_Set_Tree_Node:
+        # find the set x belongs to (with path-compression)
+        if self.map[x] != self.map[x].parent:
+            self.map[x].parent = self.find_set(self.map[x].parent.key)
+        return self.map[x].parent
```

Comment:
```suggestion
        self_map_x = self.map[x] 
        if self_map_x != self_map_x.parent:
            self_map_x.parent = self.find_set(self_map_x.parent.key)
        return self_map_x.parent
```
___OPTIONAL___: You might gain a bit of performance...
https://wiki.python.org/moin/PythonSpeed/PerformanceTips#Loops

Reply:
I'm keeping it as list comprehension, converting
```python
[disjoint_set.make_set(node) for node in self.connections]
````
to
```python
map(disjoint_set.make_set, self.connections.keys())
````
yields:
```python
**********************************************************************
File "d:/Desktop/Projects/GitUploads/Python/graphs/minimum_spanning_tree_kruskal2.py", line 77, in __main__.GraphUndirectedWeighted.kruskal
Failed example:
    mst = graph.kruskal()
Exception raised:
    Traceback (most recent call last):
      File "C:\Program Files\Python37\lib\doctest.py", line 1329, in __run
        compileflags, 1), test.globs)
      File "<doctest __main__.GraphUndirectedWeighted.kruskal[7]>", line 1, in <module>
        mst = graph.kruskal()
      File "d:/Desktop/Projects/GitUploads/Python/graphs/minimum_spanning_tree_kruskal2.py", line 100, in kruskal
        parentu = disjoint_set.find_set(u)
      File "d:/Desktop/Projects/GitUploads/Python/graphs/minimum_spanning_tree_kruskal2.py", line 24, in find_set
        elem_ref = self.map[x]
    KeyError: 1
**********************************************************************
File "d:/Desktop/Projects/GitUploads/Python/graphs/minimum_spanning_tree_kruskal2.py", line 78, in __main__.GraphUndirectedWeighted.kruskal
Failed example:
    assert 5 not in mst.connections[3]
Exception raised:
    Traceback (most recent call last):
      File "C:\Program Files\Python37\lib\doctest.py", line 1329, in __run
        compileflags, 1), test.globs)
      File "<doctest __main__.GraphUndirectedWeighted.kruskal[8]>", line 1, in <module>
        assert 5 not in mst.connections[3]
    NameError: name 'mst' is not defined
**********************************************************************
1 items had failures:
   2 of   9 in __main__.GraphUndirectedWeighted.kruskal
***Test Failed*** 2 failures.
```